### PR TITLE
feat: v1.5.2 routing depth

### PR DESCRIPTION
**Pull request by MitulShah1**
## Summary

Routing depth for the `v1.5.x` line: every routed surface ranks through one strategy implementation, and the routing layer gains the operator controls comparable gateways expose — per-attempt timeouts, `429` cooldown, typed context-length failover, sticky hashing, rule target chains, bounded predicates, attribution headers, and a host-supplied price catalog. Everything ships in OSS; no config field is removed and the Go API only grows.

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] New provider
- [ ] New plugin
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Performance improvement
- [x] Refactor / code quality
- [x] Documentation / comments only
- [ ] CI / tooling

## Related issues

Part of the `v1.5.x` routing line (#422 shipped `v1.5.1`). #244 is deliberately deferred; #351 is left open — both recorded in the release ledger.

## Changes

- One ranker: embeddings, images, rerank, moderation, transcription and speech resolve their order through the same strategy as chat, built per surface so a target that cannot serve a surface is not a candidate there. `gateway_surface_ranking.go` is gone; a parity suite holds the eight surfaces to one order, zero-weight and unpriced targets included.
- `least-latency`: samples keyed by target and upstream model, five-minute TTL, time to first chunk for streams, a 10% exploration share so a leader cannot lock in; injectable clock.
- `cost-optimized`: input **plus** output price (the request's completion ceiling or 256 tokens), catalog price per model mode on every surface, equal-cost tie-break by `targets[].weight`.
- Attribution headers on every routed surface: `X-Gateway-Provider`, `X-Gateway-Target`, `X-Gateway-Model`, `X-Gateway-Attempts`; written before a stream's first chunk; `aigateway.WithRoutingAttribution` for embedders. `ferro.routing.attempt` is emitted on the request span.
- `aigateway.WithCatalog(models.Catalog)`: a host-owned price catalog replaces the embedded/remote load and refresh.
- Validation: `ab_variants[].label` required; `single` with several targets warns at load.
- `targets[].timeout` bounds one physical attempt inside `request_timeout` (streams: until the provider answers).
- A target that answers `429` is parked for its `Retry-After` (5 s default, 60 s cap), process-local, breaker untouched.
- Typed context-length failover: the provider error keeps its envelope `code`/`type`; `core.IsContextLengthError` recognises the OpenAI-compatible, Anthropic and Gemini envelopes on `400`/`413`; every other 4xx still stops.
- `strategy.sticky: { on: user, ttl: "1h" }` for `loadbalance` and `ab-test`: a stateless hash of the request `user`.
- `target_keys: [a, b]` chains on `conditions[]` and `content_conditions[]` with a hard boundary; `target_key` stays as the one-entry form.
- Conditional predicates `user`, `stream`, `has_tools`, and `metadata` + `field` reading the single `X-Gateway-Metadata` header (≤ 32 scalars, ≤ 4 KiB; never forwarded).
- `strategy.failover_on_status_codes` (refuses `400/401/403/404/422`).
- Embedded dashboard: strategy panel shows `model_map`, rule chains and predicates.
- CHANGELOG `[1.5.2]`, ROADMAP, `config.example.yaml`, `config/README.md`, `internal/strategies/README.md`, `AGENTS.md` updated with each item.

## Testing

- [x] `make test` passes (unit tests, race detector) — root package included
- [x] `make lint` passes (0 issues)
- [x] New tests added for the change — a RED test per item: ranking parity suite, fake-clock latency expiry/exploration/cooldown, per-target timeout, chain boundary cases (retry-before-advance, no escape, exactness, skip open/parked, deterministic 4xx stop, cancellation stop, no mid-stream failover), context-length fixtures for three families plus near-miss 400s, sticky pin/TTL, attribution header per surface, validation tables
- [x] `make test-e2e-strategies` — 33/33 (cells sized for cooldown and the exploration share)
- [x] `npm run check --prefix web` equivalents (vitest, tsc, eslint) pass
- [x] `gorelease -base=v1.5.1`: additions only, except the comparability note below
- [x] Automated review against `main` and the review threads on this PR addressed
- [ ] `make test-integration` (not run locally)
- [ ] Manually tested against a live provider (n/a — no provider change)

### CODE REVIEW

File: gateway_surface.go [hunk 1]
```diff
@@ -940,9 +957,80 @@ func (g *Gateway) routeTranscription(ctx context.Context, req providers.Transcri
...
+			}
+		}
+		g.mu.Unlock()
+		return []string{key}, nil
+	}
+	strategy, err := g.strategyFor(surface)
+	if err != nil {
+		return nil, err
+	}
+	return strategy.SelectTargets(providers.Request{Model: model})
```
Comment (MitulShah1) [inline gateway_surface.go]: **Bug: non-chat surfaces silently lose `strategy.sticky` (on `user`) and `conditional`'s `user` predicate.**

`surfaceTargetOrder` builds a synthetic request carrying only `Model` before calling `SelectTargets`:

```go
return strategy.SelectTargets(providers.Request{Model: model})
```

Chat and streaming don't do this — `routeChat` (`gateway_pipeline.go`) and `streamingTargetOrder` (`gateway_stream_start.go`) both pass the real `req` into `SelectTargets`, so `req.User` is available there. But `EmbeddingRequest` and `ImageRequest` genuinely carry a `User` field (`providers/core/embedding.go:10`, `providers/core/image.go:89`), and it's dropped here before it ever reaches the strategy.

Two features this PR adds read `req.User`:
- `LoadBalance.SelectTargets` / `ABTest.SelectTargets` → `lb.sticky.unit(req.User)` (`internal/strategies/loadbalance.go:58`, `internal/strategies/abtest.go:131`) — with `req.User == ""`, `Sticky.enabled() && key != ""` is always false, so sticky routing on `user` silently never pins an embeddings/images caller to a target.
- `Conditional.SelectTargets` → `req.User != "" && req.User == rule.Value` (`internal/strategies/conditional.go:80`) — always false for these surfaces, so a `key: user` rule can never match an embeddings/image request; it always falls through to the fallback target.

Concretely: `strategy.mode: loadbalance` with `sticky: {on: user}` (or `mode: conditional` with a `key: user` rule), two embedding-capable targets — a client repeatedly calling `/v1/embeddings` with `user: "alice"` is *not* pinned/matched the way the identical config correctly pins/matches on `/v1/chat/completions`, because chat passes the real request through and this path doesn't.

Suggest passing the real request (or at least `Model` + `User`) through here, e.g. `strategy.SelectTargets(providers.Request{Model: model, User: req.User})`, plumbed from each `routeXxx` caller. Worth a ranking-parity-style test alongside the existing sticky/conditional suites, since none currently exercise this path.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by MitulShah1 — COMMENTED** [review]
## Review summary

Reviewed the diff against `main` (routing pipeline, new strategy features — sticky hashing, rule chains, cost/latency ranking, cooldown parking — config validation, the context-length/error-envelope classifier, and the HTTP handler/attribution wiring).

**One confirmed functional bug**, left as an inline comment on `gateway_surface.go:1010`: `surfaceTargetOrder` (used by embeddings, images, rerank, moderations, transcription, and speech) resolves strategy target order from a synthetic `providers.Request{Model: model}` that drops `User` (and other request fields), while chat/streaming pass the real request through. That means this PR's new `strategy.sticky` (`on: user`) and `conditional`'s `user` predicate work correctly for chat but are silently inert for every non-chat surface, even though `EmbeddingRequest`/`ImageRequest` genuinely carry a `User` field. No existing test (ranking-parity, sticky, or surface tests) currently exercises this path.

Everything else I traced — the sticky-key plumbing into chat/streaming, the rule-chain walk and its interaction with circuit breakers/cooldown parking, `EffectiveMaxTokens` resolution, the context-length classifier, and config validation coverage for the new strategy fields — looked consistent with the documented design and appropriately test-covered.

**Recommendation: request changes** — the inline finding is a real regression in newly-added routing behavior for six of the gateway's routed surfaces and looks straightforward to fix (thread `req.User`, or the full request, into `surfaceTargetOrder`'s `SelectTargets` call) plus worth a regression test alongside the existing sticky/conditional suites.

Note: this was a single-pass review of a large diff (~4600 lines across 79 files) rather than an exhaustive audit, so it's worth a second look for anything outside the areas I focused on above.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: gateway_pipeline.go [hunk 2]
```diff
@@ -349,9 +378,17 @@ func (g *Gateway) healthyKeys(plan targetPlan) []string {
...
 	defer g.mu.RUnlock()
 
+	now := g.now()
 	var healthy []string
 	for i, key := range plan.keys {
+		open := false
 		if cb := g.circuitBreakers[key]; cb != nil && cb.State() == circuitbreaker.StateOpen {
+			open = true
+		}
+		if until, parked := g.cooldowns[key]; parked && until.After(now) {
```
[the remark this answers is not in the document]
Comment (MitulShah1) [reply gateway_pipeline.go:388]: Kept as is, deliberately. A park filters the way an open circuit does and never refuses a request outright: when every target that serves the model is parked, the walk still attempts one and the upstream 429 reaches the caller, exactly as the all-circuits-open case answers 503 rather than 404. Excluding the last parked target would turn "everything is throttled" into a refusal without an attempt. This is documented in AGENTS.md and the strategies README and pinned by TestPipeline_EveryTargetParkedStillAttempts; 49ea7dbd additionally makes the filter fail open when the only healthy sibling cannot serve the model.
